lecture_notes/2017-02-14.md

```diff
@@ -0,0 +1,362 @@
+  and [PLDI 2002](./resources/fd-pldi-2002.pdf) papers.
+
+> M: (about Rust's keynote) you know what Landin said: as soon as the
+>    theoreticians get ahold of an idea, it's all downhill.
+
+> J: why did Cyclone fail?
+> D: good question
+> J: maybe **the** question (perhaps, something that cannot be overcome)
+> D: well Rust credits Cyclone as a major influence
+> M: academic projects don't last. That's it.
+
+### Fundamentals
+
+#### Sequent Calculus
+
+Girard introduced linear logic in a 1987 paper.
+Interestingly, there's a remark in the paper that lienar logic could be used
+to speed up computations.
+
+> W: Did Girard even know what that sentence meant? M., do you want to chime in?
+> M: no comment
+> D: anyway, this is the paper people go back to
+> M: well that's because Girard was engaged with the CS community. Much more
+>    than any other logician.
+> N: hmm, wasn't the linear logic paper published in a CS journal? (yes, TCS)
+
+Linear logic's roots are in Gentzen's 1934 work on the sequent calculus.
+People definitely realized the sequence calculus had impliciations for programming;
+also people were connecting logic to programming earlier in the 1930s.
+
+> M: when we give credit to Kolmogorov etc. for connecting logic to programming,
+>    I think we are a little too generous because they did not explicitly
+>    connect logic to work on partial recursive functions. That was a non-trivial
+>    step.
+
+> X: wait! you have to talk about the subformula property
+
+Right, here's the cut rule (with both propositions and language terms):
+
+```
+  Γ ⊢ t : A   Γ,x:A ⊢ u : B
+  -------------------------
+  Γ ⊢ u{t/x} : B
+```
+
+Gentzen's "key lemma" (German: Hauptsatz) is that any proof using the cut rule
+can be translated to a proof that doesn't use the cut rule. This is a key lemma
+because:
+- with it, Gentzen could prove that the sequent calculus is consistent
+  (by showing there is no way to prove a contradiction without using cuts)
+- cut elimination implies a _subformula property_ --- that one can build a
+  larger proof from smaller proofs, just as one can build a large program by
+  composing functions
+
+#### Linear Logic
+
+Linear logic is similar to the sequent calculus.
+
+Biggest difference is the absence of structural rules:
+- no _contraction_, cannot duplicate assumptions in the context
+- no _weakening_, cannot discard assumptions in the context
+
+When proving `Γ ⊢ A`, every fact in `Γ` must be used _exactly once_.
+Consequently, the sequent calculus rules for products no longer work out (would
+need to discard/duplicate propositions). Instead, there are 2 kinds of products
+in linear logic:
+- "tensor" (`A ⊗ B`) : to eliminate a tensor, must use both facts at once
+- "with" (`A & B`) : can project either conjunct, but to introduce a "with" need
+  to prove both sides simultaneously
+
+Girard also has an "of course" operator (`A!`). Contraction and weaking holds
```

Comment:
`!A` rather than `A!` I think... or at least in the early presentations it was always that way.